### DISCUSSION TITLE
[prebuild] Android packagingOptions

### DISCUSSION THREAD

**Discussion by EvanBacon** [expo/expo#16244] [Proposals and Ideas]
# Problem

Packages like `react-native-pdf`, `react-native-reanimated`, `react-native-v8`, Google Cast, anything that requires C++ really, all require custom `packagingOptions` in the `app/build.gradle`. 

To avoid dangerous regexes I created a (currently undocumented) set of static keys in the `gradle.properties` which are read and applied by the template: [PR](https://github.com/expo/expo/pull/15863). The drawback with this mechanism is that it requires [parsing the properties](https://github.com/expo/expo-cli/blob/b470e90/packages/config-plugins/src/android/Properties.ts), splitting the value by the arbitrary delimiter (`,`) and then performing an array operation, all of which needs to be done in an Expo config plugin.

I propose we expose something like `expo.android.packagingOptions` or `expo.android.build.packagingOptions` in the `app.json` which would contain the static `string[]` values:

```json
{
  "packagingOptions": {
    "pickFirsts": [],
    "excludes": [],
    "merges": [],
    "doNotStrip": []
  }
}
```

These would then be applied safely inside of the default configs in `@expo/prebuild-config`, similar to how we apply the `android.jsEngine` property.

## Drawbacks

- Setting a bigger standard for having build.gradle dynamic values being managed by the Expo config.
- No mechanism for removing values, only for adding them. Or rather, the mechanism changes to being fully in JSON.
- This would potentially break a bunch of dangerous mods (tolerable).
- Adding another field that only works in EAS Build and not in the legacy Expo Build Service.

# Notice

Since proposing this, Android has [deprecated the API](https://developer.android.com/reference/tools/gradle-api/7.0/com/android/build/api/dsl/PackagingOptions#recommendations-link) in favor of an incubating API (as is standard for Google).

The new API uses `packagingOptions.jniLibs` and `packagingOptions.resources`, each with their own packaging options. It's unclear if we want to introduce this many new properties to the `app.json` at this time.

**Comment by ide** [comment]
Looking at the drawbacks:

1. Setting aside any implementation details like having a separate parser for gradle.properties, where do you think these packaging options belong? What other code is affected by these properties? What code needs to change when these properties change, and vice versa? Group together related concepts and the lines of separation will follow.
2. Does the current gradle.properties implementation allow for removing values (looking at the linked PR, it seems the answer is no)?
3. It seems we could address breaking changes smoothly. Deprecate and warn in the next major release, and then remove and break in the major release after that.
4. Classic Builds are complete decommissioned on January 4, 2023. At most we may want to add a warning if someone runs `expo build:*` or triggers a classic build some other way on a project that specifies packagingOptions.

**Reply by EvanBacon** [reply]
We should be fine on 3. since the mods are clearly marked as dangerous. The current implementation does not allow for removing properties, but that API isn't documented yet so I'm just propagating the drawback to this API.
